pkg/resource/api.go

```diff
@@ -203,7 +202,10 @@ func (a *APIUpdatingApplicator) Apply(ctx context.Context, o runtime.Object, ao
 		}
 	}
 
-	return errors.Wrap(a.client.Update(ctx, current), "cannot update object")
+	// NOTE(hasheddan): we must set the resource version of the desired object
+	// to that of the current or the update will always fail.
+	o.(metav1.Object).SetResourceVersion(current.(metav1.Object).GetResourceVersion())
```

Comment:
In practice this type assertion should not be an issue. However, the `runtime.Object` interface is not a superset of `metav1.Object`, so maybe we should check this here?

Reply:
I think that's fine. I don't really expect a case where an object is `runtime.Object` but not `metav1.Object`.